pkg/kernel/tools/heal/liveness_check.go

```diff
@@ -40,51 +41,80 @@ func KernelLivenessCheck(deadlineCtx context.Context, conn *networkservice.Conne
 		log.FromContext(deadlineCtx).Warnf("ping is not supported for mechanism %v", mechanism)
 		return true
 	}
+	ipContext := conn.GetContext().GetIpContext()
+	addrCount := len(ipContext.GetDstIpAddrs()) * len(ipContext.GetSrcIpAddrs())
+	if addrCount == 0 {
+		log.FromContext(deadlineCtx).Debug("No IP address")
+		return true
+	}
 
 	deadline, ok := deadlineCtx.Deadline()
 	if !ok {
 		deadline = time.Now().Add(defaultTimeout)
 	}
+	timeout := time.Until(deadline)
 
-	addrCount := len(conn.GetContext().GetIpContext().GetDstIpAddrs())
-	if addrCount == 0 {
-		log.FromContext(deadlineCtx).Debug("No dst IP address")
-		return true
-	}
-	timeout := time.Until(deadline) / time.Duration(addrCount)
+	// Start ping for all Src/DstIPs combination
+	responseCh := make(chan error, addrCount)
+	defer close(responseCh)
+	for _, srcIPNet := range ipContext.GetSrcIPNets() {
+		for _, dstIPNet := range ipContext.GetDstIPNets() {
+			// Skip if IPs don't belong to the same family
+			if (srcIPNet.IP.To4() != nil) != (dstIPNet.IP.To4() != nil) {
+				responseCh <- nil
+				continue
+			}
 
-	var pinger *ping.Pinger
+			go func(srcIP, dstIP string) {
+				logger := log.FromContext(deadlineCtx).WithField("srcIP", srcIP).WithField("dstIP", dstIP)
+				pinger, err := ping.NewPinger(dstIP)
+				if err != nil {
+					logger.Errorf("Failed to create pinger: %s", err.Error())
+					responseCh <- err
+					return
+				}
+				pinger.Source = srcIP
+				pinger.SetPrivileged(true)
+				pinger.Interval = timeout / packetCount
+				pinger.Timeout = timeout
+				pinger.Count = packetCount
 
-	for _, cidr := range conn.GetContext().GetIpContext().GetDstIpAddrs() {
-		addr, _, err := net.ParseCIDR(cidr)
-		if err != nil {
-			log.FromContext(deadlineCtx).Errorf("ParseCIDR failed: %s", err.Error())
-			return false
-		}
+				err = pinger.Run()
+				if err != nil {
+					logger.Errorf("Ping failed: %s", err.Error())
+					responseCh <- err
+					return
+				}
 
-		ipAddr := &net.IPAddr{IP: addr}
-		if pinger == nil {
-			pinger, err = ping.NewPinger(addr.String())
-			if err != nil {
-				log.FromContext(deadlineCtx).Errorf("Failed to create pinger: %s", err.Error())
-				return false
-			}
-			pinger.SetPrivileged(true)
-			pinger.Interval = timeout / packetCount
-			pinger.Timeout = timeout
-			pinger.Count = packetCount
-		} else {
-			pinger.SetIPAddr(ipAddr)
-		}
-		err = pinger.Run()
-		if err != nil {
-			log.FromContext(deadlineCtx).Errorf("Ping failed: %s", err.Error())
-			return false
+				if pinger.Statistics().PacketsRecv == 0 {
+					err = errors.New("No packets received")
+					logger.Errorf(err.Error())
+					responseCh <- err
+					return
+				}
+				responseCh <- nil
+			}(srcIPNet.IP.String(), dstIPNet.IP.String())
```

Comment:
I feel without good test cases this code could produce new bugs on ci.

Can we add a few unit tests?

For this reason we could use interface/option for pinger.